consumer/consumererror/scrapeerrors.go

```diff
@@ -0,0 +1,75 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package consumererror
+
+import (
+	"fmt"
+	"strings"
+)
+
+// ScrapeErrors contains multiple PartialScrapeErrors and can also contain generic errors.
+type ScrapeErrors struct {
```

Comment:
What about you count the `failedScrapeCount` directly every time Add or Addf is called, then the Combine becomes:

```
// Combine converts a slice of errors into one error.
// It will return a PartialScrapeError if at least one error in the slice is a PartialScrapeError.
func (s *ScrapeErrors) Combine() error {
	if s.scrapeErrsCount == 0 {
		return CombineErrors(s.errs)
	}
	return NewPartialScrapeError(CombineErrors(s.errs), s.failedScrapeCount)
}
```

Am I missing something?

Reply:
The problem is that we can't assume that `ScrapeErrors` struct doesn't contain `PartialScrapeError` if `s.scrapeErrsCount == 0` since we sometimes add `PartialScrapeError` with 0 as a value for `Failed` field:

https://github.com/open-telemetry/opentelemetry-collector/pull/2414/files#diff-b1f35e860f01ab3c22cee011a3c2dd39230aa49a8e6fc041b34343cba8f231e9R159
https://github.com/open-telemetry/opentelemetry-collector/pull/2414/files#diff-b1f35e860f01ab3c22cee011a3c2dd39230aa49a8e6fc041b34343cba8f231e9R164

This creates problems (at least with tests). I updated PR so it can be shown in CI tests runs.

I guess it can be fixed by extending `consumererror.CombineErrors` to check if provided errs slice contains `PartialScrapeError`.

Reply:
I decided to leave that check in `ScrapeErrors .Combine()` for now.

Reply:
If the worry is about 0 meaning that partial errors are still there, we can have a bool as well. But that is an optimization we can do later.